### ISSUE TITLE
Support `constexpr` logic in `__attribute__(( naked ))` functions

### ISSUE THREAD

**Issue by agatha2** [llvm/llvm-project#84075]
Summary: Clang does not compile `__attribute__(( naked ))` functions containing `constexpr` control flow, despite the fact that this is sensible. Suggest adding support for that.

---

Consider the following example:
```cpp
//(Compile with "-masm=intel")

template<bool use_cmov>
__attribute__(( naked ))
uint32_t bitscan_test( uint32_t val )
{
	if constexpr (use_cmov)
	{
		asm(
			"bsf     eax, edi\n"
			"cmovz   eax, %0\n"
			"ret\n"
			: : "rm"(32) : "edi", "cc"
		);
	}
	else
	{
		asm(
			"test   edi, edi\n"
			"je     .ZERO\n"

			"bsf    eax, edi\n"
			"ret\n"

			".ZERO:\n"
			"mov    eax, 32\n"
			"ret\n"

			: : : "eax", "edi", "cc"
		);
	}
}
```
This code selects an implementation to test based on the template argument. The function is declared `naked` because we want to use `ret` directly (and we want to do that to test the branching behavior in the second implementation).

GCC compiles this just fine, but Clang produces:
```text
test.cpp:<line>:2: error: non-ASM statement in naked function is not supported
        if constexpr (use_cmov)
        ^
```
This error seems to have been added to catch cases where the user added normal C / C++ code in the function. Such code could, but probably doesn't, make sense. However, `constexpr` control flow isn't really code at all.

In this context, I read Clang's message of "not supported" as "not valid" instead of "not implemented". However, it is definitely valid, and it *could* be implemented. This issue suggests that it should be.

**Comment by shafik** [comment]
Looks like it rejects anything not `AsmStmt` or `NullStmt`:

https://github.com/llvm/llvm-project/blob/d32bcbf6a7f5beb63ce435c2bea737d33d5b7468/clang/lib/Sema/SemaDecl.cpp#L16383-L16388

I am not sure how much work is actually involved, maybe naively just allowing `IfStmt` and checking `isConstexpr()` is sufficient but perhaps there are more consequences and work to be done. 

CC @erichkeane

**Comment by erichkeane** [comment]
> Looks like it rejects anything not `AsmStmt` or `NullStmt`:
> 
> https://github.com/llvm/llvm-project/blob/d32bcbf6a7f5beb63ce435c2bea737d33d5b7468/clang/lib/Sema/SemaDecl.cpp#L16383-L16388
> 
> I am not sure how much work is actually involved, maybe naively just allowing `IfStmt` and checking `isConstexpr()` is sufficient but perhaps there are more consequences and work to be done.
> 
> CC @erichkeane

Yeah, there is a lot of consequences here that we would need to deal with.  The big part of `naked` is that it not do anything with registers (other than in ASM), since prologue and epilogues are omitted.  This is an old attribute that I presume was put in place for 'enough GCC compat to work for normal workloads', and hasn't been touched since.

We could PERHAPS allow some code, but I'm not sure of the value compared to the level of care or implementation effort it would take to do this right.  We effectively can allow any compiletime-constant code, but filtering out to that (particularly when the rule is so trivial) is not particularly easy.

So this comes down to for me, "patches welcome", but "also make sure you've done all the work to make sure no side effects can happen in the C code".

**Comment by agatha2** [comment]
The `naked` specifier is useful for bare assembly routines, like the above example, or e.g. for interrupt handlers (although there is `__attribute__((interrupt))` instead). In short: it's useful for instruction-level optimization and tinkering, and OS development.

Putting C / C++ code alongside the ASM is definitely a can of worms. However, this report is about compile-time logic, not runtime logic. The code is compile-time, and so definitionally *already* does not touch registers, because it does not run at all. It ought to be as simple as just compiling any remaining `asm` blocks.

**Comment by erichkeane** [comment]
> The `naked` specifier is useful for bare assembly routines, like the above example, or e.g. for interrupt handlers (although there is `__attribute__((interrupt))` instead). In short: it's useful for instruction-level optimization and tinkering, and OS development.
> 
> Putting C / C++ code alongside the ASM is definitely a can of worms. However, this report is about compile-time logic, not runtime logic. The code is compile-time, and so definitionally _already_ does not touch registers, because it does not run at all. It ought to be as simple as just compiling any remaining `asm` blocks.

The way `if constexpr` is implemented is a bit more complicated than that, and our runtime vs compiletime generation isn't really well split up.  So this DOES end up being a very sizable effort to make the differentiation properly.
